src/Backend/ReasonML.idr

```diff
@@ -0,0 +1,82 @@
+import Backend.Utils
+import Types
+import Typedefs
+
+%default partial
+%access public export
+
+formatVar : String -> String
+formatVar = (strCons '\'') . lowercase
+
+-- TODO would be nice if type was `(e : Env n) -> Vect (fst (Vect.filter isLeft e)) String`, 
+-- since it then could be used by several backends. I'm having trouble getting it to typecheck.
+getFreeVars : Env n -> String
+getFreeVars e = withSep ", " (either formatVar (const "")) $ snd $ Vect.filter isLeft e
+
+makeType : Env n -> TDef n -> String
+makeType     _ T0            = "void"
+makeType     _ T1            = "unit"
+makeType {n} e (TSum xs)     = tsum xs
+  where
+  tsum : Vect (2 + _) (TDef n) -> String
+  tsum [x, y]              = "either" ++ (parens $ (makeType e x) ++ ", " ++ makeType e y)
+  tsum (x :: y :: z :: zs) = "either" ++ (parens $ (makeType e x) ++ ", " ++ tsum (y :: z :: zs))
+makeType     e (TProd xs)    = assert_total $ parens $ withSep ", " (makeType e) xs
+makeType     e (TVar v)      = either formatVar lowercase $ Vect.index v e
+makeType     e (TMu nam _)   = let freeVars = getFreeVars e
+                                in if freeVars == "" then lowercase nam else lowercase nam ++ parens freeVars
+makeType     _ (TName nam _) = lowercase nam
+
+makeDefs : Env n -> TDef n -> State (List Name) String
+makeDefs _ T0           = pure ""
+makeDefs _ T1           = pure ""
+makeDefs e (TProd xs)   = map concat $ traverse (makeDefs e) xs
+makeDefs e (TSum xs)    = map concat $ traverse (makeDefs e) xs
+makeDefs _ (TVar v)     = pure ""
+makeDefs e (TMu nam cs) = 
+  do st <- get 
+     if List.elem nam st then pure "" 
+      else let
+         freeVars = getFreeVars e
+         dataName = if freeVars == "" then nam else nam ++ parens freeVars
+         newEnv = Right (if freeVars == "" then dataName else dataName) :: e
+         args = withSep " | " (mkArg newEnv) cs
+        in
+       do res <- map concat $ traverse {b=String} (\(_, bdy) => makeDefs newEnv bdy) cs 
+          put (nam :: st)
+          pure $ res ++ "\ntype " ++ lowercase dataName ++ " = " ++ args ++ ";\n"
+  where
+  mkArg : Env (S n) -> (Name, TDef (S n)) -> String
+  mkArg _ (cname, T1)       = cname
+  mkArg e (cname, TProd xs) = uppercase cname ++ parens (withSep ", " (makeType e) xs)
+  mkArg e (cname, ctype)    = uppercase cname ++ parens (makeType e ctype)
+makeDefs e (TName nam body) = 
+  do st <- get 
+     if List.elem nam st then pure "" 
+       else let 
+          freeVars = getFreeVars e
+          typeName = if freeVars == "" then lowercase nam else lowercase nam ++ parens freeVars
+         in
+        do res <- makeDefs e body 
+           put (nam :: st)
+           pure $ res ++ "\ntype " ++ typeName ++ " = " ++ makeType e body ++ ";\n"
+  
+
+-- generate type body, only useful for anonymous tdefs (i.e. without wrapping Mu/Name)
+generateType : TDef n -> String
+generateType {n} = makeType (freshEnv n)
+
+-- generate data definitions
+generate : TDef n -> String
```

Comment:
More of a general question: would it be possible to start formulating some sort of interface for functions such as this one?

Reply:
Yes I think we should, will start trying things out!

Reply:
The problem is that because we're generating a simple `String`, there's really nothing we can capture about the target language on the type level, unless we start generating syntax trees or something similar. (Which seems like it could be a good idea, but a slightly bigger decision than I'm comfortable making still...)

Reply:
Actually, it seems like a good start to say which `-> String` functions we have in the first place. That would also at least give us a good place to add some doc comments. Maybe some of them should be `-> Maybe String` instead? I'm not sure. Anyway, if we want to get more detailed at some point, we can always do that.